### coda

The tiny core of the coda system is contained in the base module.  To run this demo, your PYTHONPATH should include the .../py directory of coda. 

In [1]:
from base import *  # tiny core system

By definition, data is a finite sequence of data.  This may sound circular, but it isn't.  It's **almost** circular.  From the definition, the empty sequence of data is data.  An empty data is created in python by `data()`.  This can be displayed either as `()` or as a symbol representing the empty set `0` in python by using `repr(data())` or `str(data())` respectively.

In [2]:
#
#  create an empty data and display it
#
z = data()
print(repr(z),z)

() 0


The data constructor takes tuples of data as arguments. 

In [3]:
zz = data(z,z,z,z)
print(repr(zz),zz)

(()()()()) (0000)


In addition to the data constructor, there are two foundational operations from data to data. If x and y are data, the operations are

1. Concatenation, indicated by x y in the language or by x+y in python.
2. Colon, indicated by x:y in the language or by x|y in python. 

As you can see, all data are merely recursive sequences of empty sequences.

In [4]:
L = [z,z|z,z|(z|z),data(z|z,z|z,(z|(z|z)))]
for l in L: print(repr(l))

()
(())
(()())
((())(())(()()))


In [5]:
#
#   z|z is defined as a zero bit and 
#   z|(z|z) is defined as a one bit.
#
for l in L: print(repr(l),' -> ',str(l))

()  ->  0
(())  ->  𝟎
(()())  ->  𝟏
((())(())(()()))  ->  (𝟎𝟎𝟏)


For more examples, we can use the Universe module to create all data with specified width and depth less than specified values. The 3,2 universe, for example is all data with width <= 3 and depth <= 2.  

In [6]:
#
#   The 1,1 universe contains only empty and the zero bit. 
#
import Universe
for d in Universe.universe(1,1): print(repr(d),' -> ',d)

()  ->  0
(())  ->  𝟎


In [7]:
#
#   The 2,1 universe also contains binary 1. 
#
for d in Universe.universe(2,1): print(repr(d),' -> ',d)

()  ->  0
(())  ->  𝟎
(()())  ->  𝟏


In [8]:
U = Universe.universe(3,2)
for d in U: print(d)

0
𝟎
𝟏
(000)
(𝟎)
(𝟏)
((000))
(𝟎𝟎)
(𝟎𝟏)
(𝟎(000))
(𝟏𝟎)
(𝟏𝟏)
(𝟏(000))
((000)𝟎)
((000)𝟏)
((000)(000))
(𝟎𝟎𝟎)
(𝟎𝟎𝟏)
(𝟎𝟎(000))
(𝟎𝟏𝟎)
(𝟎𝟏𝟏)
(𝟎𝟏(000))
(𝟎(000)𝟎)
(𝟎(000)𝟏)
(𝟎(000)(000))
(𝟏𝟎𝟎)
(𝟏𝟎𝟏)
(𝟏𝟎(000))
(𝟏𝟏𝟎)
(𝟏𝟏𝟏)
(𝟏𝟏(000))
(𝟏(000)𝟎)
(𝟏(000)𝟏)
(𝟏(000)(000))
((000)𝟎𝟎)
((000)𝟎𝟏)
((000)𝟎(000))
((000)𝟏𝟎)
((000)𝟏𝟏)
((000)𝟏(000))
((000)(000)𝟎)
((000)(000)𝟏)
((000)(000)(000))


In [9]:
#
#   The sizes of universes grow enormously rapidly.
#
print(len(Universe.universe(3,3)))

60922


## Universe Sizes for given widths and depths 

| Depth     | 1  |  2  |  3    |  4  |      5  |
|----------:|---:|----:|------:|----:|--------:|
| Width 1:   | 2  |  3  |  4    |  5  |      6  |
| Width 2:   | 3  |  9  | 51    | 1857| 3265299 | 
| Width 3:   | 4  | 43  | 60922 | 
| Width 4:   | 5  | 345 | 
| Width 5:   | 6  | 3911| 
| Width 6:   | 7  | 55993 |
| Width 7:   | 8  | 960807 | 

In [10]:
U = Universe.universe(7,2)
print(len(U))

960807
